Rewrite scope completion 2

techee opened the pull request:
These are the updated patches from the original "scope completion rewrite" pull request (I wanted to preserve all the review comments so pushing separately). I hope I haven't forgotten anything from the review (but there were quite many comments so no guarantees).

There are a few more patches, mainly from the following two categories:
1. elimination of anon_struct_* and similar if possible
2. making non-scope completion more efficient by using only binary search (eliminating the linear part) and getting only as many tags as we need for the popup.

kugel- commented:
Doesn't merge cleanly with 1.25

techee commented:
Thanks for noticing, should be fixed now (I've force pushed as I hope nobody is reviewing it now and there are also no comments to lose).

techee commented:
@b4n Please let me know when you are finished with the review - I'll update the patches based on your comments afterwards.

b4n commented:
@techee I think I done reading the patches.  I merged it in a local branch to test it for a while, but please, go ahead and fix things.

One thing that these changes seem to have broken is completing members of a static structrue like this:
```c
static struct {
	int a, b;
} foo;

int main(void) {
	return foo.a;  // <- here nothing pops up after the dot -- yet used to
}
```
I didn't investigate why yet (maybe because anonymous structs are skipped somewhere they shouldn't be?).

Also, although this one didn't work before either, it'd be nice to skip array indexing too when searching what to complete while at it, so stuff like that would work:
```c
typedef struct {
	int el1, el2;
} Test;

Test test[2] = {
	{ 1, 2 },
	{ 3, 4 },
};

int main(void) {
	return test[0].el1; // <- here
}
```

kugel- commented:
Awesome, please merge. l'll try to give it another test too

elextr commented:
@kugel- slow down, @techee has to make the changes from the latest comments :)

techee commented:
Alright, I've repushed the anonymous struct handling fix - having the global counter for anonymous types caused a problem with unit tests. With the global per-all-file counter the anonymous struct name would become different every time and the generated tags files might be different every time the tag file is generated.

I hope the current approach works well - all this namespace search code is like walking through a minefield - you never know what gets broken if you make a change.

techee commented:
Forgot to mention, this isn't meant to be the final pull request - I'd just like some feedback on the last four patches and then make the remaining minor corrections (I'll probably make a new pull request for the final version).

techee commented:
@b4n If you have some time, would you have a look at the last four patches if they look OK to you? I would then create the (hopefully) final pull request addressing your comments. It would be nice to get rid of this beast finally.

b4n commented on src/editor.c:
```diff
@@ -697,51 +700,111 @@ static void request_reshowing_calltip(SCNotification *nt)
+	else if (ft->id == GEANY_FILETYPES_C || ft->id == GEANY_FILETYPES_CPP ||
+		ft->id == GEANY_FILETYPES_PHP || ft->id == GEANY_FILETYPES_RUST)
+	{
+		if (match_last_chars(sci, pos, "::"))
```
just a thought: what about using `tm_tag_context_separator()` instead of checking for some filetypes?  Not sure it's sensible at all, but right now I see the check above will leave out GLSL, Zephir and PowerShell (not sure if it'd make sense for the latter, but definitely for the former two).

techee replied:
Good idea. Even though for ASCIIDOC and TXT2TAGS it returns some dummy character (that nobody will ever enter so we don't have to worry), for the rest of the languages it really seems to return the scope separator used in the language itself. Will change it this way.

b4n commented:
@techee done reading new patches, and apart from the comments, look good.  I imported the changes in my local Geany so I'll be testing it a bit, but for now it looks good :)

b4n commented:
Just a thought (I don't think it has anything to do directly with the changes here, but it's related to scope completion): it'd be nice to trigger scope completion back when continuing on typing on a word after a scope element, instead of only when actually *typing* the scope separator.  E.g the user could cancel the autoc some way, and get back to it without re-typing the separator.

```c
static struct {
    int first, second;
} foo;

int main(void) {
    return foo.f/*cursor*/;  // <- here, typing "o" would show "first"
}
```

techee commented:
> @techee done reading new patches, and apart from the comments, look good. I imported the changes in my local Geany so I'll be testing it a bit, but for now it looks good :)

@b4n Fantastic! Thanks a lot Colomban and sorry for the huge and complicated patch sets like this - will write just trivial patches from now on :-)

I'll post the hopefully final pull request soon.

techee commented:
@b4n I agree with the scope completion popup improvement - I'll have a look at it later. For now I created a new issue here not to forget:

https://github.com/geany/geany/issues/861

kugel- commented:
Fwiw, I've been using this patch for a while and haven't noticed any problem so far
